core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsProvider.java

```diff
@@ -0,0 +1,202 @@
+
+    /**
+     * @see #builder()
+     */
+    private ContainerCredentialsProvider(Builder builder) {
+        this.credentialsEndpointProvider = builder.credentialsEndpointProvider;
+        this.credentialsFetcher = new HttpCredentialsProvider(
+            Validate.notNull(builder.credentialsEndpointProvider, "credentialsEndpointProvider"),
+            builder.asyncCredentialUpdateEnabled,
+            "elastic-container-credentials-provider");
+    }
+
+    /**
+     * Create a builder for creating a {@link ContainerCredentialsProvider}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public AwsCredentials getCredentials() {
+        if (!credentialsEndpointProvider.isEnabled()) {
+            throw new SdkClientException(credentialsEndpointProvider.disabledExceptionMessage());
+        }
+        return credentialsFetcher.getCredentials();
+    }
+
+    @Override
+    public void close() {
+        credentialsFetcher.close();
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName();
+    }
+
+    /**
+     * Retrieve the endpoint that should be used to communicate with the elastic container service to retrieve credentials.
+     */
+    public static class EcsCredentialsEndpointProvider implements CredentialsEndpointProvider {
+        @Override
+        public URI endpoint() {
+            return invokeSafely(() -> new URI(AwsSystemSetting.AWS_CONTAINER_SERVICE_ENDPOINT.getStringValueOrThrow() +
+                           AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_RELATIVE_URI.getStringValueOrThrow()));
+        }
+
+        @Override
+        public CredentialsEndpointRetryPolicy retryPolicy() {
+            return new ContainerCredentialsRetryPolicy();
+        }
+
+        @Override
+        public boolean isEnabled() {
+            return AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_RELATIVE_URI.getStringValue().isPresent();
+        }
+
+        @Override
+        public String disabledExceptionMessage() {
+            return String.format(
+                "Credentials cannot be loaded from ECS because the ECS credentials environment variable (%s) and system "
+                + "property (%s) are not set or cannot be accessed due to the security manager.",
+                AwsSystemSetting.AWS_CONTAINER_SERVICE_ENDPOINT.environmentVariable(),
+                AwsSystemSetting.AWS_CONTAINER_SERVICE_ENDPOINT.property());
+        }
+    }
+
+    public static class FullUriGenericCredentialsEndpointProvider implements CredentialsEndpointProvider {
+
+        private static Set<String> ALLOWED_HOSTS =
+            Collections.unmodifiableSet(new HashSet<>(Arrays.asList("localhost", "127.0.0.1")));
+
+        @Override
+        public URI endpoint() throws IOException {
+            URI uri = invokeSafely(() -> new URI(AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.getStringValueOrThrow()));
```

Comment:
We could use URI.create here to avoid the invokeSafely call.

Reply:
Good call.